### ISSUE TITLE
v3~beta3: opam repo suggestions

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/opam#3012]
In the context of [this discussion](https://discuss.ocaml.org/t/searching-for-a-manual-about-packaging-compiler-for-opam-2-0). I wanted to make a few suggestions: 

1. About the idea of moving things to `opam switch repo`. I'm not sure it makes sense: a lot of other `opam` commands would also move there otherwise.
2. It seems to me that one of the confusing thing is the conflation of creation/deletion to the global list with adding/removing to switches. I understand the need to have a small command set but conflating too much into subcommands makes things less clear to the end user. 
3. I think that `--this-switch` is a bad name, IIRC correctly the terminology is either `--default-switch` or `--current-switch`. One should be picked and made consistent everywhere in `opam`.

Regarding 2., I really think that adding/removing from the global list should be kept separate from adding/removing to switches. The command set make clear what is acting on switches and what is acting on repos (the manpage aswell). More precisely here's a first try at the kind of synopsis I would see:

```
# Handling global repo list
repo new REPO [--default] [ADDRESS] [QUORUM] [FINGERPRINTS]
repo delete REPO
repo set REPO KEY VALUE  # With KEY (default|url|quorum|fingerprint)
repo get REPO KEY

# Handling repo and switches
# Note that repo lists below REPO[,REPO]... have to be specified as a single arg.
# If no switch is explicitly specified 
repo add REPO[,REPO]... [SWITCH]...
repo remove REPO[,REPO]... [SWITCH]...
repo list [--all] [--default] [SWITCH]...
repo set-priority REPO RANK [SWITCH]...
repo set-priority-list REPO[,REPO]... [SWITCH]... 
```

This would simplify the `SCOPE SPECIFICATION OPTIONS` which I would call `SWITCH SELECTION OPTIONS` to 
```
-a, --all-switches 
--current-switch  # Default if no positional argument is specified
[SWITCH]...       # Positional arguments
```

If you really want to keep `add` + `new` workflow (but personally I don't mind) I would add the following options usable on `repo add`:

```
--new=URL
--new-quorum=QUORUM
--new-fingerprint=FINGERPRINT 
```

However I'm not necessarily happy about the `new/delete`, `add/remove` and the `list` option which is still confusing the global list with the repo switch list. Besides `new/delete` is a bit inconsistent with `switch create/remove`, `pin add/remove`. 

Maybe we could try to slightly tweak the terminology and talk about *attaching* repos to switches. This would lead to the following synopsis:

```
# Handling the global list
repo add REPO [--default] [ADDRESS] [QUORUM] [FINGERPRINTS]
repo remove REPO
repo list [--default]
repo set REPO KEY VALUE  # With KEY (default|url|quorum|fingerprint)
repo get REPO KEY

# Handling repo and switches
repo attach REPO[,REPO]... [SWITCH]...
repo detach REPO[,REPO]... [SWITCH]...
repo attached [SWITCH]...
repo set-priority REPO RANK [SWITCH]...
repo set-priority-list REPO[,REPO]... [SWITCH]... 
```

**Comment by dbuenzli** [comment]
Note that `set-priority-list` is a bit different from the existing `set-repo` (which somehow can be seen as `attach`, even though `set-repo` would delete I guess). 

I think that `set-priority-list` should enforce the order give on the cli while prerserving as much as possible the previous order.

**Comment by AltGr** [comment]
I agree it's confusing as it is. My first prototype had `add`/`remove` and `select`/`unselect`, with a meaning similar to what you propose last (with `attach`/`detach`). It was even more confusing, but maybe because of bad defaults and options.

I like your first proposition better: it's functionally equivalent to what I propose [here](https://discuss.ocaml.org/t/searching-for-a-manual-about-packaging-compiler-for-opam-2-0/638/6?u=altgr), but the different names may be a good idea. Having both `delete` and `remove` is clearly a problem, though, as you point out.

There may still be a way with just `add`/`remove`, but similar semantics, if they just manage the global list (don't affect any switches or the default setting by default), unless explicitely with the "switch selection options".

```
# Handling the global list
repo add REPO ADDRESS [QUORUM] [FINGERPRINTS]    # advertise attaching the repo
repo remove REPO  # this becomes a global remove, ask for confirmation
repo list # list all, highlights the ones currently attached
...

# Handling repo and switches
repo add REPO [ADDRESS] [QUORUM] [FINGERPRINTS] --current-switch|--switch SW|--default
repo remove REPO --current-switch|--switch SW|--default
repo list --current-switch|--switch SW|--default
...
```
Note that `repo add REPO` without either address or switch arguments would be an error, and `repo add REPO ADDRESS` is an error if `REPO` exists and has a different address.

**Comment by dbuenzli** [comment]
I still think that overloading is `add/remove` is really not a good idea. Somehow the fact that `--current-switch`, `--switch SW` and `--default` options do not represent the same entities is a bit suspicious to me. 

Note that both proposition are the same except for the names but I really think the first one is undesirable. I don't see exactly what is confusing with `attach/detach` (maybe the *select* vocabulary was confusing). Introducing the attachment vocabulary will be useful for discussing repositories in relationship to switches vs repositories in relation to the global list.
 
Also my proposal is able to handle both simple attach/detach requests: one repo to a single/current switch and complex ones, attach/detach many repos to many switches with a simple cli (no need to specify options) which I see as a good sign

Finally if you don't want `attached` you could keep a little bit of overloading on `list [--default-repos|--all-repos] [SWITCH]...`.

**Comment by AltGr** [comment]
FWIW, linking the original conversation about this: https://github.com/ocaml/opam/pull/2663#r78400661

**Comment by dbuenzli** [comment]
> FWIW, linking the original conversation about this: #2663 (comment)

I think that this discussion misses the following points:

1. I suspect it was trying to get rid of the notion of `select`ion which is something in UI that I see as rather transient: a set of objects to apply an operation on. It was trying to eliminate vocabulary that maybe should have been there but with a different name.
2. It failed to see that lists of repos (`all`, `default`, switch specific) are more complex beasts than suggested there: `quorum`, `fingerprint`, `default` and `priority`. Not all these attributes are relevant for each list so having a unifying command set with list selection via option argument is not necessarily the best design. Having a clear delienation in the command set to operate on the different lists may be easier to understand and enable a better UI.

**Comment by dbuenzli** [comment]
> repo attached [SWITCH]...

This could actually be `opam switch repos`.

**Comment by dbuenzli** [comment]
I just went to the process of defining a repo with a new compiler package and I think that conflating the command set for the global list and the list of repos attached to a switch really doesn't work well. 

I added my repo without the `--set-default` flag which means that it wouldn't show up in the list of available switches and the only way I found to make it show up there was to remove the repo with `--all` and then start over by specifying this option.

I think the best would be to distinguish the command set and allow, in the command set for attaching switches, to specify an address in which case an interactive mode allows you to add to the global list.
